### PR TITLE
motif: fix build with clang

### LINKED ISSUE #302449: Build failure: motif
### Steps To Reproduce

1. build motif

### Build log

output of `nix log /nix/store/7m27hlg1wdymiqf22drbdwcqh7jk9j5d-motif-2.3.8.drv | sed  -e 's/\x1b\[[0-9;]*m//g'` (to remove color codes)
https://gist.github.com/cbarrett/302fb8409beaa8a36a8206ce9c77ba0a

### Additional context

I'm on Mac OS 12.7.2, with Xcode 14.2. FWIW my ultimate goal was to build snd, which is how I noticed motif failing.

### Notify maintainers

@alyssais @trofi

### Metadata

```console
~ $ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 21.6.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `""`
 - channels(cbarrett): `"darwin, nixpkgs-23.11-darwin"`
 - nixpkgs: `/Users/cbarrett/.nix-defexpr/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by alyssais**
## Description of changes

Fixes https://github.com/NixOS/nixpkgs/issues/302449.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by alyssais** [comment]
@OfBorg eval

**Review by paparodeo — COMMENTED** [review]
darwin has wchar.h and wide character string functions so removing code in an ifdef with (broken) definitions will fix the build. one of the below  fixes will work. both tested on x64 darwin using current master.

<details><summary>replace-APPLE.diff</summary>

```diff
diff --git a/pkgs/development/libraries/motif/default.nix b/pkgs/development/libraries/motif/default.nix
index 31d4ba6385c1..6e05bb3c5afd 100644
--- a/pkgs/development/libraries/motif/default.nix
+++ b/pkgs/development/libraries/motif/default.nix
@@ -56,6 +56,12 @@ stdenv.mkDerivation rec {
     })
   ];
 
+  # darwin has wchar.h and wide char string functions so no need for the
+  # (broken) workaround
+  postPatch = ''
+    substituteInPlace lib/Xm/TextF.c --replace-fail 'defined(__APPLE__)' '0'
+  '';
+
   enableParallelBuilding = true;
 
   meta = with lib; {
```
</details>

alternatively, if a patch is prefered:
<details><summary>patch-out-APPLE.diff</summary>

```diff
diff --git a/pkgs/development/libraries/motif/TextF-fix-darwin.diff b/pkgs/development/libraries/motif/TextF-fix-darwin.diff
new file mode 100644
index 000000000000..a7a0a1dffe18
--- /dev/null
+++ b/pkgs/development/libraries/motif/TextF-fix-darwin.diff
@@ -0,0 +1,13 @@
+diff --git a/lib/Xm/TextF.c b/lib/Xm/TextF.c
+index f66076d..6496911 100644
+--- a/lib/Xm/TextF.c
++++ b/lib/Xm/TextF.c
+@@ -82,7 +82,7 @@ static char rcsid[] = "$TOG: TextF.c /main/65 1999/09/01 17:28:48 mgreess $"
+ #define FIX_1409
+ 
+ #if (defined(__FreeBSD__) && (__FreeBSD__ < 4)) || \
+-    (defined(__APPLE__) || defined(__NetBSD__) || defined(__OpenBSD__))
++    (defined(__NetBSD__) || defined(__OpenBSD__))
+ /*
+  * Modification by Integrated Computer Solutions, Inc.  May 2000
+  *
diff --git a/pkgs/development/libraries/motif/default.nix b/pkgs/development/libraries/motif/default.nix
index 31d4ba6385c1..9087764fb08b 100644
--- a/pkgs/development/libraries/motif/default.nix
+++ b/pkgs/development/libraries/motif/default.nix
@@ -33,6 +33,7 @@ stdenv.mkDerivation rec {
   patches = [
     ./Remove-unsupported-weak-refs-on-darwin.patch
     ./Add-X.Org-to-bindings-file.patch
+    ./TextF-fix-darwin.diff
     (fetchpatch rec {
       name = "fix-format-security.patch";
       url = "https://raw.githubusercontent.com/void-linux/void-packages/b9a1110dabb01c052dadc1abae1413bd4afe3652/srcpkgs/motif/patches/02-${name}";
```
</details>

**Comment by alyssais** [comment]
Thanks a lot for looking into it!  It looks like MacPorts has the same patch, so we can fetch it from there:

https://github.com/macports/macports-ports/blob/master/x11/openmotif/files/wcs-functions.patch

**Review by paparodeo — APPROVED** [review]
Result of `nixpkgs-review` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>cernlib</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>xbill</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>catclock</li>
    <li>motif</li>
    <li>nedit</li>
    <li>snd</li>
  </ul>
</details>

xbill is unrelated. broken due to configure script etc.

File: pkgs/development/libraries/motif/default.nix [hunk 1]
```diff
@@ -27,7 +27,7 @@ stdenv.mkDerivation rec {
   propagatedBuildInputs = [ libXp libXau ];
 
   prePatch = lib.optionalString (!demoSupport) ''
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/motif/default.nix:29]: Shouldn't this really be postPatch so that patches apply properly?

Comment (alyssais) [reply pkgs/development/libraries/motif/default.nix:29]: I suppose so!

Comment (alyssais) [reply pkgs/development/libraries/motif/default.nix]: Changed.

**Comment by paparodeo** [comment]
ZHF: https://github.com/NixOS/nixpkgs/issues/309482

**Pull request merged**